README.md
# coast

![build](https://github.com/spkersten/flutter_coast/actions/workflows/master.yml/badge.svg?branch=master)

PageView wrapper supporting Hero-like animations

![img/example](img/example.gif)

## Example

Create a `Coast` with a number of `Beaches` and add a `CrabController` as observer
to the `Coast`. Wrap a widget that you want to animate between beaches with a
`Crab` widget with the same `tag`.

```dart
class CoastExampleApp extends StatefulWidget {
  @override
  _CoastExampleAppState createState() => _CoastExampleAppState();
}

class _CoastExampleAppState extends State<CoastExampleApp> {
  final _beaches = [
    Beach(builder: (context) => Cadzand()),
    Beach(builder: (context) => Westkapelle()),
    Beach(builder: (context) => Zoutelande()),
  ];

  final _coastController = CoastController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Coast(
        beaches: _beaches,
        controller: _coastController,
        observers: [
          CrabController(),
        ],
      ),
    );
  }
}


class Cadzand extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(
      title: Text("Cadzand"),
      backgroundColor: Colors.deepOrange,
    ),
    body: Stack(
      children: [
        Positioned(
          top: 40,
          left: 40,
          child: Crab(
            tag: "container",
            child: Container(
              color: Colors.green,
              width: 80,
              height: 60,
            ),
          ),
        ),
      ],
    ),
  );
}

class Westkapelle extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(
      title: Text("Westkapelle"),
      backgroundColor: Colors.amber,
    ),
    body: Stack(
      children: [
        Positioned(
          top: 80,
          right: 40,
          child: Crab(
            tag: "container",
            child: Container(
              color: Colors.green,
              width: 200,
              height: 100,
            ),
          ),
        ),
      ],
    ),
  );
}

class Zoutelande extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
    appBar: AppBar(
      title: Text("Zoutelande"),
      backgroundColor: Colors.deepPurple,
    ),
    body: Stack(
      children: [
        Positioned(
          bottom: 40,
          left: 20,
          child: Crab(
            tag: "container",
            child: Container(
              color: Colors.green,
              width: 100,
              height: 200,
            ),
          ),
        ),
      ],
    ),
  );
}
```

CHANGELOG.md
## 2.0.2

* Make CoastController listenable

## 2.0.1

* Allow settings the initial page

## 2.0.0

* Null-safety

## 1.0.2

* Remove accidentally exposed page snapping property

## 1.0.1

* Extended the README

## 1.0.0

* Initial release


test/coast_test.dart
import 'dart:async';

import 'package:coast/coast.dart';
import 'package:coast/src/coast.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_test_ui/flutter_test_ui.dart';

void main() {
  group('$Coast', () {
    group('two beaches, one crab', () {
      const startRect = Rect.fromLTWH(0, 0, 40, 40);
      const endRect = Rect.fromLTWH(400 - 80, 400 - 80, 80, 80);

      late CoastController coastController;

      setUpUI((tester) async {
        await tester.binding.setSurfaceSize(const Size(400, 400));

        coastController = CoastController();

        await tester.pumpWidget(MaterialApp(
          home: Coast(
            controller: coastController,
            observers: [
              CrabController(),
            ],
            beaches: [
              Beach(
                builder: (context) => Align(
                  alignment: Alignment.topLeft,
                  child: Crab(
                    tag: 'crab',
                    child: Container(color: Colors.red, width: 40, height: 40),
                  ),
                ),
              ),
              Beach(
                builder: (context) => Align(
                  alignment: Alignment.bottomRight,
                  child: Crab(
                    tag: 'crab',
                    child: Container(color: Colors.red, width: 80, height: 80),
                  ),
                ),
              ),
            ],
          ),
        ));
      });

      testUI('initially at the first beach, with the crab at the starting position', (tester) async {
        expect(coastController.beach, 0.0);

        expect(
          tester.getRect(find.byType(Container)),
          startRect,
        );
      });

      group('drag halfway to the second beach', () {
        late TestGesture drag;

        setUpUI((tester) async {
          drag = await tester.startGesture(const Offset(200, 200));
          await drag.moveBy(const Offset(-100, 0));
          await tester.pump();
          await drag.moveBy(const Offset(-100, 0));
          await tester.pump();
        });

        testUI('the crab is halfway between the start and end position', (tester) async {
          expect(coastController.beach, 0.5);

          expect(
            tester.getRect(find.byType(Container)),
            Rect.lerp(startRect, endRect, 0.5),
          );
        });

        group('let the coast snap to the second beach', () {
          setUpUI((tester) async {
            await drag.moveBy(const Offset(-100, 0));
            await drag.up();
            await tester.pumpAndSettle();
          });

          testUI('the crab is at the final position', (tester) async {
            expect(coastController.beach, 1);

            expect(
              tester.getRect(find.byType(Container)),
              endRect,
            );

            expect(find.byType(Container), findsOneWidget);
          });
        });

        group('let the coast snap back to the first beach', () {
          setUpUI((tester) async {
            await drag.moveBy(const Offset(100, 0));
            await drag.up();
            await tester.pumpAndSettle();
          });

          testUI('the crab is at the final position', (tester) async {
            expect(coastController.beach, 0);

            expect(
              tester.getRect(find.byType(Container)),
              startRect,
            );

            expect(find.byType(Container), findsOneWidget);
          });
        });
      });
    });
  });

  group('$CoastState', () {
    group('calculate NewSourcePage', () {
      final sut = CoastState();

      test('offset bit bigger then source page should return current source page', () {
        expect(sut.calculateNewSourcePage(offset: 4.6, sourcePage: 4), 4);
      });

      test('offset much bigger then source page should recalculate source page', () {
        expect(sut.calculateNewSourcePage(offset: 4.9, sourcePage: 2), 4);
      });

      test('offset much smaller then source page should calculate source page based on offset', () {
        expect(sut.calculateNewSourcePage(offset: 1.9, sourcePage: 3), 2);
      });

      test('offset bit smaller then source page should return current source page', () {
        expect(sut.calculateNewSourcePage(offset: 2.3, sourcePage: 3), 3);
      });
    });

    group('calculate NewTargetPage', () {
      final sut = CoastState();

      test('offset is bigger then source page and should determine new target page correctly', () {
        expect(sut.calculateNewTargetPage(offset: 4.6, newSourcePage: 3), 4);
      });

      test('offset is smaller then source page and should determine new target page correctly', () {
        expect(sut.calculateNewTargetPage(offset: 2.6, newSourcePage: 3), 2);
      });

      test('return null when offset equals sourcePage', () {
        expect(sut.calculateNewTargetPage(offset: 3, newSourcePage: 3), null);
      });
    });

    group('finish transition', () {
      final sut = CoastState();

      test('should finish transition in case new target page is not equal to current target page', () {
        sut.progress = TransitionAnimation();
        expect(sut.shouldFinishTransition(newTargetPage: 1, newSourcePage: null), true);
      });

      test('should finish transition in case new source page is not equal to current source page', () {
        sut.progress = TransitionAnimation();
        expect(sut.shouldFinishTransition(newTargetPage: null, newSourcePage: 1), true);
      });

      test('should not finish transition in case ', () {
        sut.progress = null;
        expect(sut.shouldFinishTransition(newTargetPage: null, newSourcePage: null), false);
      });
    });

    group('start new transition', () {
      final sut = CoastState();

      test('Should not start transition when target page and progress is null', () {
        sut.progress = null;
        expect(sut.shouldStartNewTransition(newTargetPage: null), false);
      });

      test('should start new transition in case target page has value', () {
        sut.progress = null;
        expect(sut.shouldStartNewTransition(newTargetPage: 2), true);
      });

      test('should start not start new transition in case progress has value', () {
        sut.progress = TransitionAnimation();
        expect(sut.shouldStartNewTransition(newTargetPage: null), false);
      });
    });
  });

  group('TransitionAnimation', () {
    group('Calculate animation status', () {
      final sut = TransitionAnimation();

      test('should return dismissed in case value 0', () {
        expect(sut.calculateStatus(0), AnimationStatus.dismissed);
      });

      test('should return completed in case value 1', () {
        expect(sut.calculateStatus(1), AnimationStatus.completed);
      });

      test('should return forward in case newvalue is bigger then previous value', () {
        sut.value = 0.5;

        expect(sut.calculateStatus(0.7), AnimationStatus.forward);
      });

      test('should return reverse in case newvalue is bigger then previous value', () {
        sut.value = 0.5;

        expect(sut.calculateStatus(0.3), AnimationStatus.reverse);
      });
    });

    group('Event handling', () {
      final updates = <AnimationStatus>[];
      late StreamController<AnimationStatus> animationUpdates;
      late TransitionAnimation sut;

      setUp(() async {
        animationUpdates = StreamController<AnimationStatus>(sync: true);
        animationUpdates.stream.listen(updates.add);
        sut = TransitionAnimation()..addStatusListener(animationUpdates.add);
      });

      tearDown(() async {
        updates.clear();
        await animationUpdates.close();
      });

      test('Should notify listener only once with new status', () async {
        sut.value = 1.0;

        expect(updates.length, 1);
      });

      test('Should notify listener on each value change', () async {
        sut
          ..value = 1.0
          ..value = 0.5;

        expect(updates.length, 2);
      });

      test('Should dispatch correct status update object to listener', () async {
        sut.value = 1.0;

        expect(updates.first, AnimationStatus.completed);
      });
    });
  });
}


lib/src/crab.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

import 'coast.dart';

/// Analogue of [Hero] but for [Coast] and [Beach]'s instead of for [Navigator] and [Route]'s. The implementation is
/// mostly adapted for [Hero] and related classes.
class Crab extends StatefulWidget {
  const Crab({
    required this.tag,
    required this.child,
    this.placeholderBuilder,
    this.flightShuttleBuilder,
    Key? key,
  }) : super(key: key);

  final String tag;
  final Widget child;
  final TransitionBuilder? placeholderBuilder;
  final CrabFlightShuttleBuilder? flightShuttleBuilder;

  @override
  State<StatefulWidget> createState() => _CrabState();

  static Map<Object, _CrabState> _allCrabsFor(BuildContext context) {
    final result = <Object, _CrabState>{};
    void visitor(Element element) {
      if (element.widget is Crab) {
        final crab = element as StatefulElement;
        final crabWidget = element.widget as Crab;
        final Object tag = crabWidget.tag;
        assert(() {
          if (result.containsKey(tag)) {
            throw FlutterError('There are multiple crabs that share the same tag within a subtree.\n'
                'Within each subtree for which carbs are to be animated (typically a Beach subtree), '
                'each Crab must have a unique non-null tag.\n'
                'In this case, multiple carbs had the following tag: $tag\n'
                'Here is the subtree for one of the offending carbs:\n'
                '${element.toStringDeep(prefixLineOne: '# ')}');
          }
          return true;
        }());
        final crabState = crab.state as _CrabState;
        result[tag] = crabState;
      }
      // Don't perform transitions across different Coasts.
      if (element.widget is Navigator) {
        return;
      }
      element.visitChildren(visitor);
    }

    context.visitChildElements(visitor);
    return result;
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(StringProperty('tag', tag))
      ..add(ObjectFlagProperty<TransitionBuilder?>.has('placeholderBuilder', placeholderBuilder))
      ..add(ObjectFlagProperty<CrabFlightShuttleBuilder?>.has('flightShuttleBuilder', flightShuttleBuilder));
  }
}

class _CrabState extends State<Crab> {
  final GlobalKey _key = GlobalKey();
  Size? _placeholderSize;

  void startWalk() {
    assert(mounted);
    final box = context.findRenderObject() as RenderBox?;
    assert(box != null && box.hasSize);
    setState(() {
      _placeholderSize = box!.size;
    });
  }

  void endWalk() {
    if (mounted) {
      setState(() {
        _placeholderSize = null;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_placeholderSize != null) {
      if (widget.placeholderBuilder == null) {
        return SizedBox(width: _placeholderSize!.width, height: _placeholderSize!.height);
      } else {
        return widget.placeholderBuilder!(context, widget.child);
      }
    }
    return KeyedSubtree(
      key: _key,
      child: widget.child,
    );
  }
}

/// A function that lets [Crab]s self supply a [Widget] that is shown during the
/// crabs's walk from one beach to another instead of default (which is to
/// show the destination beach's instance of the Crab).
typedef CrabFlightShuttleBuilder = Widget Function(
  BuildContext flightContext,
  Animation<double> animation,
  BeachTransitionDirection direction,
  BuildContext fromCrabContext,
  BuildContext? toCrabContext,
);

/// Adapted from [HeroController], animates matching [Crab]'s when transitioning between [Beach]es.
class CrabController extends CoastObserver {
  CrabController() : _walks = {};

  final Map<Object, _CrabWalk> _walks;

  @override
  void startTransition(
    Beach beach,
    Beach previousBeach,
    BeachTransitionDirection direction,
    Animation<double> progress,
  ) {
    if (previousBeach.subtreeContext == null) {
      return;
    }

    final fromCrabs = Crab._allCrabsFor(previousBeach.subtreeContext!);

    final overlay = coast.overlay!;
    final overlayRect = _globalBoundingBoxFor(overlay.context);

    /// Because [beach] hasn't been laid out yet, we don't know the rects of that Crabs
    /// on that Beach. Therefore we have to create the crab walk in a post frame callback.
    /// To prevent the [Crab] from moving with [previousBeach] in the current frame, we
    /// start a walk for all [Crabs] to the same position on screen, keeping them stationary.
    /// In the post frame callback, we update the walk with the target rect for the Crab.
    for (final tag in fromCrabs.keys) {
      final fromShuttleBuilder = fromCrabs[tag]?.widget.flightShuttleBuilder;
      final shuttleBuilder = fromShuttleBuilder ?? _defaultHeroFlightShuttleBuilder;

      _walks[tag] = _CrabWalk(
        fromCrab: fromCrabs[tag]!,
        toCrab: null,
        progress: progress,
        overlay: overlay,
        overlayRect: overlayRect,
        direction: direction,
        fromCrabContext: previousBeach.subtreeContext!,
        toCrabContext: beach.subtreeContext,
        shuttleBuilder: shuttleBuilder,
      )..start();
    }

    WidgetsBinding.instance!.addPostFrameCallback((value) {
      final context = beach.subtreeContext;
      if (context != null) {
        final toCrabs = Crab._allCrabsFor(context);

        for (final tag in _walks.keys) {
          final toCrab = toCrabs[tag];
          if (toCrab != null) {
            _walks[tag]!.update(toCrab, context);
          }
        }
      }
    });
  }

  Widget _defaultHeroFlightShuttleBuilder(
    BuildContext flightContext,
    Animation<double> animation,
    BeachTransitionDirection flightDirection,
    BuildContext fromCrabContext,
    BuildContext? toCrabContext,
  ) {
    if (toCrabContext != null) {
      final toCrab = toCrabContext.widget as Crab;
      return toCrab.child;
    } else {
      final fromCrab = fromCrabContext.widget as Crab;
      return fromCrab.child;
    }
  }
}

class _CrabWalk {
  _CrabWalk({
    required this.fromCrab,
    required this.toCrab,
    required this.progress,
    required this.overlay,
    required this.overlayRect,
    required this.direction,
    required this.fromCrabContext,
    required this.toCrabContext,
    required this.shuttleBuilder,
  });

  final _CrabState fromCrab;
  _CrabState? toCrab;
  final Animation<double> progress;
  final OverlayState overlay;
  final Rect overlayRect;
  final BeachTransitionDirection direction;
  final BuildContext fromCrabContext;
  BuildContext? toCrabContext;
  final CrabFlightShuttleBuilder shuttleBuilder;

  OverlayEntry? _overlayEntry;

  late Tween<Rect?> _crabRectTween;

  void start() {
    progress.addStatusListener(_handleAnimationStatusUpdates);

    _crabRectTween = RectTween(
      begin: _globalBoundingBoxFor(fromCrab.context, ancestor: fromCrabContext.findRenderObject()),
      end: _globalBoundingBoxFor(fromCrab.context, ancestor: fromCrabContext.findRenderObject()),
    );

    fromCrab.startWalk();

    _overlayEntry = _createOverlayEntry();
    overlay.insert(_overlayEntry!);
  }

  void update(_CrabState newToCrab, BuildContext newToCrabContext) {
    toCrab = newToCrab;
    toCrabContext = newToCrabContext;

    toCrab!.startWalk();

    _crabRectTween = RectTween(
      begin: _globalBoundingBoxFor(fromCrab.context, ancestor: fromCrabContext.findRenderObject()),
      end: _globalBoundingBoxFor(toCrab!.context, ancestor: toCrabContext!.findRenderObject()),
    );

    _overlayEntry?.remove();
    _overlayEntry = _createOverlayEntry();
    overlay.insert(_overlayEntry!);
  }

  OverlayEntry _createOverlayEntry() => OverlayEntry(
        builder: (context) {
          final shuttle = shuttleBuilder(context, progress, direction, fromCrab.context, toCrab?.context);

          return AnimatedBuilder(
            animation: progress,
            builder: (context, child) {
              final rect = _crabRectTween.evaluate(progress)!;
              final size = overlayRect.size;
              final offsets = RelativeRect.fromSize(rect, size);

              return Positioned(
                top: offsets.top,
                left: offsets.left,
                bottom: offsets.bottom,
                right: offsets.right,
                child: IgnorePointer(
                  child: child,
                ),
              );
            },
            child: shuttle,
          );
        },
      );

  void _handleAnimationStatusUpdates(AnimationStatus status) {
    if (status == AnimationStatus.completed || status == AnimationStatus.dismissed) {
      _overlayEntry!.remove();
      _overlayEntry = null;

      fromCrab.endWalk();
      toCrab?.endWalk();
    }
  }
}

Rect _globalBoundingBoxFor(BuildContext context, {RenderObject? ancestor}) {
  final box = (context.findRenderObject() as RenderBox?)!;
  assert(box.hasSize);
  return MatrixUtils.transformRect(box.getTransformTo(ancestor), Offset.zero & box.size);
}


lib/src/coast.dart
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/widgets.dart';

/// A wrapper of [PageView] that support observation of transitions between pages.
class Coast extends StatefulWidget {
  const Coast({
    required this.beaches,
    required this.controller,
    this.observers,
    this.physics,
    this.allowImplicitScrolling = false,
    this.restorationId,
    this.scrollDirection = Axis.horizontal,
    this.reverse = false,
    this.onPageChanged,
    this.dragStartBehavior = DragStartBehavior.start,
    this.clipBehavior = Clip.hardEdge,
    Key? key,
  })  : assert(beaches.length > 0),
        super(key: key);

  final List<Beach> beaches;
  final List<CoastObserver>? observers;
  final CoastController controller;

  /// See [PageView]
  final ScrollPhysics? physics;
  final bool allowImplicitScrolling;
  final String? restorationId;
  final Axis scrollDirection;
  final bool reverse;
  final ValueChanged<int>? onPageChanged;
  final DragStartBehavior dragStartBehavior;
  final Clip clipBehavior;

  @override
  State<StatefulWidget> createState() => CoastState();

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(IterableProperty<Beach>('beaches', beaches))
      ..add(IterableProperty<CoastObserver>('observers', observers))
      ..add(DiagnosticsProperty<CoastController>('controller', controller))
      ..add(DiagnosticsProperty<ScrollPhysics?>('physics', physics))
      ..add(DiagnosticsProperty<bool>('allowImplicitScrolling', allowImplicitScrolling))
      ..add(StringProperty('restorationId', restorationId))
      ..add(EnumProperty<Axis>('scrollDirection', scrollDirection))
      ..add(DiagnosticsProperty<bool>('reverse', reverse))
      ..add(ObjectFlagProperty<ValueChanged<int>?>.has('onPageChanged', onPageChanged))
      ..add(EnumProperty<DragStartBehavior>('dragStartBehavior', dragStartBehavior))
      ..add(EnumProperty<Clip>('clipBehavior', clipBehavior));
  }
}

class CoastController extends ChangeNotifier {
  CoastController({int initialPage = 0}) : _pageController = PageController(initialPage: initialPage) {
    _pageController.addListener(_listener);
  }

  final PageController _pageController;

  double? get beach => _pageController.page;

  @override
  void dispose() {
    _pageController.dispose();
    super.dispose();
  }

  Future<void> animateTo({
    required int beach,
    Duration duration = const Duration(milliseconds: 200),
    Curve curve = Curves.fastOutSlowIn,
  }) async {
    await _pageController.animateToPage(beach, duration: duration, curve: curve);
  }

  void _listener() {
    notifyListeners();
  }
}

class CoastState extends State<Coast> {
  late double _previousOffset;
  TransitionAnimation? progress;
  int _sourcePage = 0;
  int? _targetPage;

  final _overlayKey = GlobalKey<OverlayState>(debugLabel: "CoastState's Overlay");

  OverlayState? get overlay => _overlayKey.currentState;

  PageController get _pageController => widget.controller._pageController;

  double _round(double value, int precision) {
    final f = pow(10, precision);
    return (value * f).round() / f;
  }

  @override
  void initState() {
    super.initState();

    _sourcePage = _pageController.initialPage;
    _previousOffset = 0.0;

    _pageController.addListener(() {
      // Get rid of over-scrolling
      final offset = _round(_pageController.page!.clamp(0.0, widget.beaches.length - 1).toDouble(), 6);
      if (offset == _previousOffset) {
        return;
      }

      // Determine between which two pages we are scrolling
      final newSourcePage = calculateNewSourcePage(offset: offset, sourcePage: _sourcePage);

      final newTargetPage = calculateNewTargetPage(offset: offset, newSourcePage: newSourcePage);

      if (shouldFinishTransition(newTargetPage: newTargetPage, newSourcePage: newSourcePage)) {
        // Finish previous transition
        progress
          ?..value = (newSourcePage < _sourcePage) ? 0.0 : 1.0
          ..dispose();
        progress = null;
        _targetPage = null;
      }

      _sourcePage = newSourcePage;

      if (shouldStartNewTransition(newTargetPage: newTargetPage)) {
        // Start new a transition
        _targetPage = newTargetPage;

        final direction = _targetPage! > _sourcePage ? BeachTransitionDirection.right : BeachTransitionDirection.left;

        progress = TransitionAnimation();

        for (final observer in widget.observers ?? <CoastObserver>[]) {
          observer
            ..coast = this
            ..startTransition(widget.beaches[_targetPage!], widget.beaches[_sourcePage], direction, progress!);
        }
      }

      // Update progress of the transition that is in progress
      if (progress != null) {
        if (_targetPage! > _sourcePage) {
          progress!.value = offset - _sourcePage;
        } else {
          progress!.value = _sourcePage - offset;
        }
      }

      _previousOffset = offset;
    });
  }

  @visibleForTesting
  int calculateNewSourcePage({required double offset, required int sourcePage}) {
    if (offset >= (sourcePage + 1)) {
      return sourcePage + (offset - sourcePage).floor();
    } else if (offset <= (sourcePage - 1)) {
      return sourcePage - (sourcePage - offset).floor();
    } else {
      return sourcePage;
    }
  }

  @visibleForTesting
  int? calculateNewTargetPage({required double offset, required int newSourcePage}) {
    if (offset > newSourcePage) {
      return newSourcePage + 1;
    } else if (offset < newSourcePage) {
      return newSourcePage - 1;
    } else {
      return null;
    }
  }

  @visibleForTesting
  bool shouldFinishTransition({int? newTargetPage, int? newSourcePage}) =>
      progress != null && (_targetPage != newTargetPage || _sourcePage != newSourcePage);

  @visibleForTesting
  bool shouldStartNewTransition({int? newTargetPage}) => progress == null && newTargetPage != null;

  @override
  Widget build(BuildContext context) => _DeclarativeOverlay(
        overlayKey: _overlayKey,
        child: LayoutBuilder(
          builder: (context, constraints) => SizedBox(
            /* PageView issue on flutter https://github.com/flutter/flutter/issues/24763
            Temporary fix to round the width till the issue is fixed */
            width: constraints.maxWidth.roundToDouble(),
            child: PageView(
              controller: _pageController,
              physics: widget.physics,
              allowImplicitScrolling: widget.allowImplicitScrolling,
              restorationId: widget.restorationId,
              scrollDirection: widget.scrollDirection,
              reverse: widget.reverse,
              pageSnapping: true,
              onPageChanged: widget.onPageChanged,
              dragStartBehavior: widget.dragStartBehavior,
              clipBehavior: widget.clipBehavior,
              children: widget.beaches.map((beach) => beach.build(context)).toList(),
            ),
          ),
        ),
      );

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties
      ..add(DiagnosticsProperty<TransitionAnimation?>('progress', progress))
      ..add(DiagnosticsProperty<OverlayState?>('overlay', overlay));
  }
}

class _DeclarativeOverlay extends StatefulWidget {
  const _DeclarativeOverlay({required this.child, required this.overlayKey, Key? key}) : super(key: key);

  final Widget child;
  final Key overlayKey;

  @override
  _DeclarativeOverlayState createState() => _DeclarativeOverlayState();

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<Key>('overlayKey', overlayKey));
  }
}

class _DeclarativeOverlayState extends State<_DeclarativeOverlay> {
  late OverlayEntry _childEntry;

  @override
  void initState() {
    super.initState();
    _childEntry = OverlayEntry(
      maintainState: true,
      builder: (context) => widget.child,
    );
  }

  @override
  void didUpdateWidget(_DeclarativeOverlay oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.child != widget.child) {
      _childEntry.markNeedsBuild();
    }
  }

  @override
  Widget build(BuildContext context) => Overlay(
        key: widget.overlayKey,
        initialEntries: [_childEntry],
      );
}

/// See [Coast]
class Beach {
  Beach({required this.builder});

  final WidgetBuilder builder;

  final GlobalKey _subtreeKey = GlobalKey();

  BuildContext? get subtreeContext => _subtreeKey.currentContext;

  Widget build(BuildContext context) => RepaintBoundary(key: _subtreeKey, child: Builder(builder: builder));
}

class CoastObserver {
  late CoastState coast;

  /// [progress] will animate between 0 and 1. It will always start at 0, and end either at 1 (when the transition
  /// completes) or at 0 (when the transition is cancelled).
  void startTransition(
      Beach beach, Beach previousBeach, BeachTransitionDirection direction, Animation<double> progress) {}
}

enum BeachTransitionDirection { left, right }

class TransitionAnimation extends Animation<double>
    with AnimationLocalListenersMixin, AnimationLocalStatusListenersMixin, AnimationEagerListenerMixin {
  double _value = 0;
  AnimationStatus? _status = AnimationStatus.dismissed;

  @override
  AnimationStatus get status => _status!;

  @override
  double get value => _value;

  set value(double newValue) {
    _status = calculateStatus(newValue);
    _value = newValue;
    notifyListeners();
    _checkStatusChanged();
  }

  AnimationStatus _lastReportedStatus = AnimationStatus.dismissed;

  @visibleForTesting
  AnimationStatus? calculateStatus(double newValue) {
    if (newValue == 0.0) {
      return AnimationStatus.dismissed;
    } else if (newValue == 1.0) {
      return AnimationStatus.completed;
    } else if (newValue > _value) {
      return AnimationStatus.forward;
    } else if (newValue < _value) {
      return AnimationStatus.reverse;
    } else {
      //this can never happen but is a fallback scenario
      return AnimationStatus.dismissed;
    }
  }

  void _checkStatusChanged() {
    final newStatus = status;
    if (_lastReportedStatus != newStatus) {
      _lastReportedStatus = newStatus;
      notifyStatusListeners(newStatus);
    }
  }
}


example/lib/main.dart
import 'package:coast/coast.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(CoastExampleApp());
}

class CoastExampleApp extends StatefulWidget {
  @override
  _CoastExampleAppState createState() => _CoastExampleAppState();
}

class _CoastExampleAppState extends State<CoastExampleApp> {
  final _beaches = [
    Beach(builder: (context) => Cadzand()),
    Beach(builder: (context) => Westkapelle()),
    Beach(builder: (context) => Zoutelande()),
  ];

  final _coastController = CoastController(initialPage: 1);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Coast(
        beaches: _beaches,
        controller: _coastController,
        observers: [
          CrabController(),
        ],
      ),
    );
  }
}

class Cadzand extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text("Cadzand"),
          backgroundColor: Colors.deepOrange,
        ),
        body: Stack(
          children: [
            Positioned(
              top: 40,
              left: 40,
              child: Crab(
                tag: "container",
                child: Container(
                  color: Colors.green,
                  width: 80,
                  height: 60,
                ),
              ),
            ),
          ],
        ),
      );
}

class Westkapelle extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text("Westkapelle"),
          backgroundColor: Colors.amber,
        ),
        body: Stack(
          children: [
            Positioned(
              top: 80,
              right: 40,
              child: Crab(
                tag: "container",
                child: Container(
                  color: Colors.green,
                  width: 200,
                  height: 100,
                ),
              ),
            ),
          ],
        ),
      );
}

class Zoutelande extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(
          title: Text("Zoutelande"),
          backgroundColor: Colors.deepPurple,
        ),
        body: Stack(
          children: [
            Positioned(
              bottom: 40,
              left: 20,
              child: Crab(
                tag: "container",
                child: Container(
                  color: Colors.green,
                  width: 100,
                  height: 200,
                ),
              ),
            ),
          ],
        ),
      );
}
